### ISSUE TITLE
Standardize third party libraries for generating string hash and unique ids

### ISSUE THREAD

**Issue by goplayoutside3** [zooniverse/front-end-monorepo#6493]
## Package
_Examples_
- app-project
- lib-classifier
- lib-react-components
- lib-user

## lib-classifier

- `cuid` is deprecated ([npm link](https://www.npmjs.com/package/cuid))
    - This package creates a temporary id per classification object stored in the Mobx store (temp id is not posted to Panoptes).
    - `cuid` is also used throughout the annotation code to generate an id for default annotation objects and new marks in the drawing tools.
    - Example in ClassificationStore.js
    https://github.com/zooniverse/front-end-monorepo/blob/496e8834ff78eb356228f512837703101bafc70a/packages/lib-classifier/src/store/ClassificationStore.js#L57-L59

- `hash.js` is a popular npm package last published 6 yrs ago, but it might replicate behavior available in modern browsers
    - The current session id is included with classification metadata. A new session id is returned if more than five minutes have passed since the last call to `sessionUtils.getSessionID()`. Sessions are stored in browser session storage, so they're unique to individual tabs.
    - ❓ Does the session id need to be cryptographically secure? Or just unique? 
    - This library is only used in session.js:
    https://github.com/zooniverse/front-end-monorepo/blob/496e8834ff78eb356228f512837703101bafc70a/packages/lib-classifier/src/store/utils/session.js#L12-L14

## lib-react-components

- `cuid` is deprecated. Same third party library as lib-classifier.
    - It's used to create an id in DataSeriesPlot.js which is part of the Media component (for subject preview thumbnails).
    - clipPath takes an id to work with `visx/group`:
    https://github.com/zooniverse/front-end-monorepo/blob/496e8834ff78eb356228f512837703101bafc70a/packages/lib-react-components/src/Media/components/Data/components/DataSeriesPlot/DataSeriesPlot.js#L165

## app-project

- `@sindresorhus/string-hash` is not a popular npm package. [Last published](https://www.npmjs.com/package/@sindresorhus/string-hash) 3 yrs ago.
    - It’s used in app-project’s Mobx store to create a non-cryptographic hashed string for a unique identifier attached to the browser cookie. The cookie determines whether a user has dismissed a project’s announcement banner.
    - This feature responds to the string content of `announcement` (which project teams can change any time) hence a random unique id is not attached to the cookie preference in favor of a hashed string of the announcement's content.
    - Example of `stringHash` in UI.js:
    https://github.com/zooniverse/front-end-monorepo/blob/496e8834ff78eb356228f512837703101bafc70a/packages/app-project/stores/UI.js#L60-L64

## lib-user

- `uuid`: a popular, maintained [npm package](https://www.npmjs.com/package/uuid).
    - Generates a unique id when creating a new group (Panoptes resource).
    https://github.com/zooniverse/front-end-monorepo/blob/496e8834ff78eb356228f512837703101bafc70a/packages/lib-user/src/components/MyGroups/components/GroupCreateFormContainer/GroupCreateFormContainer.js#L10-L16

## Describe the solution you'd like
- Standardize libraries for the three main use cases:
    1. Cryptographically secure hash
    2. Non-crypographically secure hash from a digested string
    3. Non-crypographically secure randomly generated id

- SHA256 is now included in current browsers via `crypto.subtle.digest`
    - https://stackoverflow.com/questions/59777670/how-can-i-hash-a-string-with-sha256
    - Could we use this in server-side rendered code? Will this work with our testing env?
- `uuid` is a maintained library. We could use it in place of `cuid` throughout lib-classifier and lib-react-components.
    - Would need to evaluate if `uuid` an over-large package for our use case and if tree-shaking is working correctly to grab only `v4`
    - Could we just use https://developer.mozilla.org/en-US/docs/Web/API/Crypto/randomUUID?

**Comment by eatyourgreens** [comment]
`string-hash` is a one-line wrapper for [`fnv1a`](https://github.com/sindresorhus/fnv1a), which is why it doesn’t get updated often.

**Comment by eatyourgreens** [comment]
There's a test for the case where the TTL has expired but it's a bad test. It tests that a stub has been called, rather than testing that the session ID has changed. 

https://github.com/zooniverse/front-end-monorepo/blob/master/packages/lib-classifier/src/store/utils/session.spec.js

Unfortunately, the classifier is full of tests that use stubs like this. The problem with them is that testing if a method was called is not the same as testing if a method behaved as expected.

**Comment by goplayoutside3** [comment]
Agreed it's a bad test. The condition `ttl < Date.now()` is never actually met because `ttl` is stored as a Date() string yet Date.now() is a number.

It's probably worth removing the "five minutes" logic from both PFE and FEM.

**Comment by eatyourgreens** [comment]
To answer your original question, though, I would expect that `getSessionId` returns different values if you call it more than 5 minutes apart. 

I don't think it matters if the value is secure. As far as I'm aware, project teams use it to group classifications by anonymous volunteers, so it only needs to be unique to a session.

**Comment by eatyourgreens** [comment]
You're right! I don't see how that code has ever worked. Which means any analysis based on PFE session IDs is basically flawed. 

Good catch!

**Comment by goplayoutside3** [comment]
Maybe 🤔 When someone looks at a classification's session id, without a signed-in user, do they assume it's a metadata value that refreshes every 5 minutes though? When I think "session id" as a web concept, I think of it as an individual browser tab id that refreshes when a new tab or window is opened.

**Comment by eatyourgreens** [comment]
It doesn't refresh every 5 minutes. It's supposed to refresh if there's no activity for 5 minutes. If you're classifying rapidly, each classification should have the same session ID, because the TTL updates every time you press Done. 

I'm sure it's been used to measure session length in papers. Off the top of my head, maybe the Microsoft experiment that I helped with, which looked at the effect of intervention messages on session length.

**Comment by eatyourgreens** [comment]
The per-tab sessions that you're referring to are recorded by Sugar, and counted on https://activity.zooniverse.org.

**Comment by goplayoutside3** [comment]
That's a good clarification about 5 minutes of inactivity and interesting connection to https://activity.zooniverse.org/.

I opened https://github.com/zooniverse/front-end-monorepo/pull/6499 for FEM and will do the same for PFE.

**Comment by eatyourgreens** [comment]
I think that session code needs to be fixed, rather than deleted, since classification sessions are used in published papers. A classification session is distinct from a browser session, in that a volunteer can keep a tab open for hours or days and have several classification sessions in one browser session.

There might even be code in https://github.com/zooniverse/data-digging that uses classification session IDs to estimate session durations. That'll be wrong now.

**Comment by goplayoutside3** [comment]
> I think that session code needs to be fixed, rather than deleted

Similar to my [comment](https://github.com/zooniverse/front-end-monorepo/pull/6504#issuecomment-2510273841) on your PR, I'm in favor of not changing the classifier's behavior. Session ids have never refreshed after 5min of activity in the PFE classifier or the FEM classifier.

Regarding data digging and publications, I'll address the potential incorrect assumptions about session ids with the zoo team and document our decision on the appropriate Issues.
